### PR TITLE
concatTextFile: init

### LINKED ISSUE #147589: networking.hostFiles not working
### Describe the bug
In my config I have `networking.hostFiles=[ ./mvn.hosts ];` and when building it gives me the following error:
```
nix log /nix/store/xsg1gi465km5l6zv0z32m8q8m137q818-hosts.drv
cat: /nix/store/1dmmrmi6yncn4j2ymmqziag8fflq8jiw-source/mvn.hosts: No such file or directory
```

However if I replace it with `networking.extraHosts = builtins.readFile ./mvn.hosts;` it builds without errors.

Even weirder, If I do `cat /nix/store/1dmmrmi6yncn4j2ymmqziag8fflq8jiw-source/mvn.hosts` it works without problems.

### Expected behavior
The configuration builds normally

### Notify maintainers
@Infinisil 

### Metadata
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.4, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20211006_53e4794`
 - nixpkgs: `/nix/store/8gnhm8ws354wzqd1mkapra14qk7bnc40-source`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
 - networking.hostFiles
```

### PR DESCRIPTION
**Pull request by pasqui23**
###### Motivation for this change
Fixes https://github.com/NixOS/nixpkgs/issues/147589

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/trivial-builders.nix [hunk 1]
```diff
@@ -322,6 +322,71 @@ rec {
...
+   *
+  */
+  concatText = name: files: concatTextFile { inherit name files; };
+
+    /*
+   * Writes a text file to nix store with and mark it as executable.
+   *
+   * Example:
+   * # Writes contents of files to /nix/store/<store path>
+   * concatdScript "my-file" [ file1 file2 ]
```
Comment (Artturin) [inline pkgs/build-support/trivial-builders.nix:384]: Typo

File: pkgs/build-support/trivial-builders.nix [hunk 2]
```diff
@@ -322,6 +322,71 @@ rec {
...
+  /* concat a list of files to the nix store.
+   * The contents of files are added to the file in the store.
+   *
+   * Examples:
+   * # Writes my-file to /nix/store/<store path>
+   * concatTextFile {
+   *   name = "my-file";
+   *   files = [ drv1 "${drv2}/path/to/file" ];
+   * }
+   * # See also the `concatdText` helper function below.
```
Comment (Artturin) [inline pkgs/build-support/trivial-builders.nix:335]: Typo

File: pkgs/build-support/trivial-builders.nix [hunk 3]
```diff
@@ -322,6 +322,71 @@ rec {
...
+    , checkPhase ? ""    # syntax checks, e.g. for scripts
+    , meta ? { }
+    }:
+    runCommandLocal name
+      { inherit files executable checkPhase meta destination; }
+      ''
+        n=$out$destination
+        mkdir -p "$(dirname "$n")"
+        cat $files > "$n"
+        touch "$n"
```
Comment (pennae) [inline pkgs/build-support/trivial-builders.nix:361]: is this necessary? the redirect will create the file anyway, and file times are erased in the store

File: pkgs/build-support/trivial-builders.nix [hunk 4]
```diff
@@ -322,6 +322,71 @@ rec {
...
+   * concatTextFile {
+   *   name = "my-file";
+   *   files = [ drv1 "${drv2}/path/to/file" ];
+   * }
+   * # See also the `concatdText` helper function below.
+   *
+   * # Writes executable my-file to /nix/store/<store path>/bin/my-file
+   * concatTextFile {
+   *   name = "my-file";
+   *   files = ''
```
Comment (pennae) [inline pkgs/build-support/trivial-builders.nix:340]: this example does not work:
```
nix-repl> :b concatTextFile {                                         
                      name = "my-file";
                      files = ''
                        Contents of File
                      '';
                      executable = true;
                      destination = "/bin/my-file";
                    }   
my-file> cat: Contents: No such file or directory
my-file> cat: of: No such file or directory
my-file> cat: File: No such file or directory
error: builder for '/nix/store/8gz8nnn2qs4w89jn4w7795mm4rpfjwnp-my-file.drv' failed with exit code 1;
       last 3 log lines:
       > cat: Contents: No such file or directory
       > cat: of: No such file or directory
       > cat: File: No such file or directory
```

File: pkgs/build-support/trivial-builders.nix [hunk 5]
```diff
@@ -322,6 +322,71 @@ rec {
...
+   *   '';
+   *   executable = true;
+   *   destination = "/bin/my-file";
+   * }
+   */
+  concatTextFile =
+    { name # the name of the derivation
+    , files
+    , executable ? false # run chmod +x ?
+    , destination ? ""   # relative path appended to $out eg "/bin/foo"
```
Comment (pennae) [inline pkgs/build-support/trivial-builders.nix:351]: the example path is not relative. not sure whether it'd be better to expect absolute paths or relative paths, but would prefer relative paths (even though it requires adding a `/` when concatenating). relative paths don't give the impression that `destination` could be a valid absolute path in the generated system.

File: pkgs/build-support/trivial-builders.nix [hunk 6]
```diff
@@ -322,6 +322,71 @@ rec {
...
+    , checkPhase ? ""    # syntax checks, e.g. for scripts
+    , meta ? { }
+    }:
+    runCommandLocal name
+      { inherit files executable checkPhase meta destination; }
+      ''
+        n=$out$destination
```
Comment (pennae) [inline pkgs/build-support/trivial-builders.nix:358]: maybe rename this to `file` or similar, and note that this variable exists for `checkPhase` to use

**Comment by pasqui23** [comment]
@pennae I wanted to give it an API similar to writeTextFile

**Comment by pennae** [comment]
huh, didn't realize writeTextFile uses paths of that form too. in that case consistency is more important.

File: pkgs/build-support/trivial-builders.nix [hunk 7]
```diff
@@ -322,6 +322,67 @@ rec {
...
+    , files
+    , executable ? false # run chmod +x ?
+    , destination ? ""   # relative path appended to $out eg "/bin/foo"
+    , checkPhase ? ""    # syntax checks, e.g. for scripts
+    , meta ? { }
+    }:
+    runCommandLocal name
+      { inherit files executable checkPhase meta destination; }
+      ''
+        file=$out$destination
```
Comment (roberth) [inline pkgs/build-support/trivial-builders.nix:356]:
```suggestion
        file=$out/${destination:-$name}
```

This is better because it makes this work:

```
runCommand "test" {} ''
cp ${concatText "foo.bar" []} .
mkdir -p $out
mv foo.bar $out/
''
```

Please add it as a test case; see `pkgs/tests/trivial-builders.nix`.

File: pkgs/build-support/trivial-builders.nix [hunk 8]
```diff
@@ -322,6 +322,67 @@ rec {
...
+
+    /*
+   * Writes a text file to nix store with and mark it as executable.
+   *
+   * Example:
+   * # Writes contents of files to /nix/store/<store path>
+   * concatScript "my-file" [ file1 file2 ]
+   *
+  */
+  concatScript = name: files: concatTextFile { inherit name files; executable=true; };
```
Comment (roberth) [inline pkgs/build-support/trivial-builders.nix:383]:
```suggestion
  concatScript = name: files: concatTextFile { inherit name files; executable = true; };
```

File: pkgs/build-support/trivial-builders.nix (hunk 7 above)
Comment (roberth) [reply pkgs/build-support/trivial-builders.nix:356]: Sorry, this is only feasible after #119942, so you can set `outPath` correctly. Carry on, we'll improve it later. Please do add a simpler test though.

**Comment by pasqui23** [comment]
@roberth that would make it inconsistent with writeTextFile

**Comment by roberth** [comment]
> @roberth that would make it inconsistent with writeTextFile

Bug for bug compatibility is not a requirement, but implementing correctly it is not feasible now without a huge effort to get the overriding right, so my suggestion can wait and be implemented for all such functions later. It'd be a slightly breaking change because of the potential for up-preferences (`../`) in the file. I'll write an issue.

**Review by roberth — CHANGES_REQUESTED** [review]
Just needs docs and tests.

Docs go in `doc/builders/trivial-builders.chapter.md`

Tests go in `pkgs/test/default.nix`

**Review by infinisil — COMMENTED** [review]
Looks alright, though an alternative route to fix the original problem would be to just do a `map (file: "${file}") cfg.hostFiles`, which explicitly copies the files into the store.

Concatenating files seems like it might be useful in other cases though, especially with shell scripts.

File: doc/builders/trivial-builders.chapter.md [hunk 9]
```diff
@@ -47,6 +47,12 @@ These functions write `text` to the Nix store. This is useful for creating scrip
 
 Many more commands wrap `writeTextFile` including `writeText`, `writeTextDir`, `writeScript`, and `writeScriptBin`. These are convenience functions over `writeTextFile`.
 
+## `concatTextFile`, `concatText`, `concatScript` {#trivial-builder-concatText}
+
+These functions concatenate `files` to the Nix store in a single file. This is useful for configuration files structured in lines of text. `concatTextFile` takes an attribute set and expects two arguments, `name` and `files`. `name` corresponds to the name used in the Nix store path. `files` will be the files to be concatenated. You can also set `executable` to true to make this file have the executable bit set.
+
+Many more commands wrap `concatTextFile` including `concatText` and`concatScript`, . These are convenience functions over `concatTextFile`.
```
Comment (roberth) [inline doc/builders/trivial-builders.chapter.md:54]:
```suggestion
`concatText` and`concatScript` wrap `concatTextFile` for convenience.
```

File: pkgs/build-support/trivial-builders.nix [hunk 10]
```diff
@@ -322,6 +322,67 @@ rec {
...
+   * # See also the `concatText` helper function below.
+   *
+   * # Writes executable my-file to /nix/store/<store path>/bin/my-file
+   * concatTextFile {
+   *   name = "my-file";
+   *   files = [ drv1 "${drv2}/path/to/file" ];
+   *   executable = true;
+   *   destination = "/bin/my-file";
+   * }
+   */
```
Comment (roberth) [inline pkgs/build-support/trivial-builders.nix:344]: Could you add or move the examples to the manual?

Comment (roberth) [reply pkgs/build-support/trivial-builders.nix:344]: Same for the other two.

File: pkgs/build-support/trivial-builders/test/writeStringReferencesToFile.nix [hunk 11]
```diff
@@ -1,9 +1,11 @@
-{ callPackage, lib, pkgs, runCommand, writeText, writeStringReferencesToFile }:
+{ callPackage, lib, pkgs, runCommand, writeText, concatText, writeStringReferencesToFile }:
```
Comment (roberth) [inline pkgs/build-support/trivial-builders/test/writeStringReferencesToFile.nix:1]: This is the writeStringReferencesToFile test.

Please create a separate file in this directory and a new attribute in `pkgs/test/default.nix`.

**Review by pennae — CHANGES_REQUESTED** [review]
the test does not run, and is not referenced in `pkgs/test/default.nix`.

**Review by pennae — APPROVED** [review]

File: pkgs/build-support/trivial-builders/test/concat-test.nix [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{ callPackage, lib, pkgs, runCommand, concatText, writeText, hello }:
+let
+  stri = writeText "pathToTest";
+  txt1 = stri "abc";
+  txt2 = stri hello;
+  res = concatText "textToTest" [ txt1 txt2 ];
+in
+runCommand "test-concatPaths" { } ''
+  diff -U3 <(cat ${txt1} ${txt2}) ${res}
```
Comment (roberth) [inline pkgs/build-support/trivial-builders/test/concat-test.nix:9]: Please add a test case for the empty list.

**Review by pennae — APPROVED** [review]

**Review by roberth — APPROVED** [review]
